Hide the host until all Shadow DOM blocking link stylesheets are loaded

tomalec opened the pull request:
Sorry for creating yet another PR for #93, again.
But I'd like to move faster with smaller steps, and still keep track of supporting `enlighted-link` and `stylesheet`s in light DOM

https://github.com/Starcounter/starcounter-include/issues/93

Done:
- [x] opt-out capabilities with `async` attribute,
- [x] Cover race condition between resources load and changing composition or content,
- [x] Prepare for default custom-element styles Web Platform feature
- [x] tests,
- [x] cross-browser support,
- [x] README notes

tomalec committed:
Hide the host until all Light or Shadow DOM blocking links are loaded,

https://github.com/Starcounter/starcounter-include/issues/93

Support light and Shadow DOM,
opt-out capabilities with `async` attribute,
Cover race condition between resources load and changing composition or content,
Prepare for default custom-element styles Web Platform feature

tomalec committed:
Remove case insensitive flag from BLOCKING_LINKS_SELECTOR

as it is not supported in Edge

tomalec committed:
Ignore ShadyCSS styles in more test cases

tomalec committed:
Test element behavior just after it was defined => upgraded,

not after imports are loaded, as this may be to late = after imported-template `stamped` content.

tomalec committed:
Add tests for FOUC solution,

https://github.com/Starcounter/starcounter-include/issues/93

tomalec committed:
Add Juicy/enlighted-link to bower dependencies,

to be able to test behavior related to it.

tomalec committed:
Limit FOUC solution to `link rel="stylesheet" in Shadow DOM only,

to move forward with smaller steps.
https://github.com/Starcounter/starcounter-include/issues/93

tomalec committed:
Resolve instantly when the stylesheet is already loaded

alshakero commented on README.md:
```diff
@@ -143,6 +144,11 @@ Name                                    | Detail                 | Description
 
+## Render-blocking links
+
+To mimic native behavior of markup for Declarative Shadow DOM and to avoid <abbr title="Flash Of Unstyled Content">FOUC</abbr> the element will "block rendering" until all `<link rel="stylesheet">`s are loaded or throw an error.
```
WOW! didn't know about `<abbr>`!

alshakero commented on README.md:
```diff
@@ -143,6 +144,11 @@ Name                                    | Detail                 | Description
+## Render-blocking links
+
+To mimic native behavior of markup for Declarative Shadow DOM and to avoid <abbr title="Flash Of Unstyled Content">FOUC</abbr> the element will "block rendering" until all `<link rel="stylesheet">`s are loaded or throw an error.
+You can opt-out by using the `async` attribute on your `<link>`. The blocked rendering is achieved by setting `visibility: hidden` on shadow host - `<strcounter-include>` element.
```
Can you specify that `async` is an empty attribute and its existence is enough without a value.

alshakero commented on starcounter-include.html:
```diff
@@ -15,15 +15,47 @@
+                return null;
+            }
+            let rejectDefer;
+            const promise = new Promise((resolveAll, rejectAll) => {
```
Since you're not rejectingAll. I would do `resolveAll =>`

alshakero commented on starcounter-include.html:
```diff
@@ -108,7 +140,8 @@
                         var mergedComposition = null;
-                        const templates = event.detail.querySelectorAll('template[is="declarative-shadow-dom"][presentation=default],template[is="declarative-shadow-dom"]:not([presentation])');
+                        const fragment = event.detail;
+                        const templates = fragment.querySelectorAll('template[is="declarative-shadow-dom"][presentation=default],template[is="declarative-shadow-dom"]:not([presentation])');
```
Maybe also turn these selectors into constants on top?

alshakero commented on starcounter-include.html:
```diff
@@ -15,15 +15,47 @@
+                })).then(resolveAll);
+                rejectDefer = rejectAll;
+            });
+            promise.reject = rejectDefer;
```
Can you rename this prop to `cancel`? Because currently it's tangled with Promise's typical `reject` while it's a bit different. Anything not `reject` would do.

tomalec replied:
The similarity to typical `Promise.reject` was on purpose, as it's somewhat common standard to call it that way for defered object:
 - https://developer.mozilla.org/en-US/docs/Mozilla/JavaScript_code_modules/Promise.jsm/Deferred#backwards_forwards_compatible
 - https://api.jquery.com/jquery.deferred/

warpech replied:
I recognize this pattern and I think that the name `.reject` is fine, because that's the name used in the pattern.

alshakero commented on starcounter-include.html:
```diff
@@ -476,13 +519,25 @@
          * @param {DocumentFragment} givenComposition
          */
         StarcounterIncludePrototype.stampComposition = function(givenComposition){
```
I would make this `async function(givenComposition) {`. Justification below.

tomalec replied:
below, meaning where?

tomalec replied:
The comment was not posted, but we discussed it in person.

tomalec replied:
I gave it a second thought. The function name and main responsibility is to "stampComposition" this stamping is always fully synchronous, that's why "stamped" (I'd rename it to "presentation-stamped") is called synchronously at the end.

We could rename it to `async stampAndLoadComposition(givenComposition: DocumentFragment): Promise | true`. But no other part of `starcounter-include` code waits for it. Event composition editor doesn't wait for it https://github.com/Starcounter/starcounter-layout-html-editor/blob/master/starcounter-layout-html-editor.html

"load composition" is the action that **may** be asynchronous. I'd trigger an `presentation-loaded` after it's done. WDYT?

warpech replied:
I think the `presentation-loaded` event will be enough for now (YAGNI).

alshakero commented on starcounter-include.html:
```diff
@@ -500,6 +555,18 @@
             console.error('clear is not yet defined!');
         }
 
+        // Desired use of upcomming Web Platform feature,
```
upcoming

alshakero commented on starcounter-include.html:
```diff
@@ -500,6 +555,18 @@
+        // so we would not have to implement all `loading-*` attributes and
+        // fallback composition logic.
+        // customElements.define('starcounter-include', StarcounterInclude,
+        //         { style: new CSSStyleSheet(`
```
man this would be awesome when it comes real.

alshakero commented on test/blocking-links/shadow-dom.html:
```diff
@@ -0,0 +1,111 @@
+                scInclude = document.createElement('starcounter-include');
+                testFixtureConatiner.appendChild(scInclude);
+                importedTemplate = scInclude.querySelector_template_for_buggy_polyfill();
+                importedTemplate.addEventListener('stamping', ()=>{
```
no `{}` needed.

tomalec replied:
I find it easier to read and maintain like `if(){`

alshakero commented on test/blocking-links/shadow-dom.html:
```diff
@@ -0,0 +1,111 @@
+                    }
+                };
+            });
+            afterEach(()=>{
```
No `{}` needed.

tomalec replied:
See https://github.com/Starcounter/starcounter-include/pull/111#discussion_r215342281

alshakero commented on test/blocking-links/shadow-dom.html:
```diff
@@ -0,0 +1,111 @@
+                        describe('after content is stamped, and all imports are loaded', ()=>{
+                            beforeEach(function (done) {
+                                setTimeout(done, 500);
+                            });
```
Could you comment we why need to wait this long. I know now, but will probably forget soon.

tomalec replied:
Is there something I already forgot except "and all imports are loaded" ?

alshakero replied:
No I think this would be enough

tomalec replied:
Then this is stated in `describe`

alshakero commented on test/mocked_sc_htmlmerger.js:
```diff
@@ -14,9 +14,10 @@ connect().use('/sc/htmlmerger', function fooMiddleware(req, res, next) {
+        let fileName = parts[1];
         content += `<imported-template-scope scope="${appName}"><template><meta itemprop="juicy-composition-scope" content="${appName}"/></template>`;
         if(fileName){
+            fileName = fileName.replace('%2F','/');
```
This replaces the first instance only. Are you sure that file name has only one `/`? And I would replace with [`path.sep`](https://nodejs.org/api/path.html#path_path_sep) to cover unix-like and windows.

warpech replied:
Perhaps use https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/decodeURIComponent?

alshakero replied:
`decodeURIComponent` is a Web API. This file runs within Node.

alshakero commented on test/mocked_sc_htmlmerger.js:
```diff
@@ -14,9 +14,10 @@ connect().use('/sc/htmlmerger', function fooMiddleware(req, res, next) {
         content += `<imported-template-scope scope="${appName}"><template><meta itemprop="juicy-composition-scope" content="${appName}"/></template>`;
         if(fileName){
+            fileName = fileName.replace('%2F','/');
             if(fs.existsSync(servingDirectory + fileName)){
```
This is usually a bad idea. I would do [`path.resolve(__dirname, 'sc', fileName)`](https://nodejs.org/api/path.html#path_path_resolve_paths).

warpech replied:
This is not Node, it is a web browser. I propose to use https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/decodeURIComponent

alshakero reviewed (changes requested):
This is a static-review (didn't run the code). I'll run it once you address my comments.

Brilliant work for real <3

tomalec committed:
Add an example with `async` attribute used on `link`,

https://github.com/Starcounter/starcounter-include/pull/111#discussion_r215257645

tomalec committed:
Add constant with default Shadow DOM presentation selector at the top of the script,

https://github.com/Starcounter/starcounter-include/pull/111#discussion_r215265122

warpech commented on starcounter-include.html:
```diff
@@ -15,15 +15,47 @@
         function warnAboutDeprecatedPartial() {
             console.warn('`partial` attribute and property are deprecated from `starcounter-include` in favour of (viewModel property or view-model attribute), they will soon be no longer be supported');
         }
+        const BLOCKING_LINKS_SELECTOR = 'link[rel="stylesheet"]:not([async])';
```
Since `async` is not a standard attribute, I propose to change this to `data-async`

tomalec replied:
I agree it's future-bullet-proof and according to spec recommendation (even though not all spec authors are sure if `data-` recommendation is a valid any longer in custom elements times). `data-` gives us guarantee that we will not get affected by updates Web Spec, if there  be native `async` link.

We do not use `data-` prefix for other elements and features.
Also it's quite verbose.

Maybe we could do `sc-async` or prefix it somehow with Starcounter-specific name and dash, which is unlikely to conflict with native. So, we will be 100% sure no other library or custom element will interfere with it.


Also maybe we could drop the attribute at all, and let app devs opt out as they do for regular `<link>`s added to `head` https://www.filamentgroup.com/lab/async-css.html
```html
<link rel="preload" href="mystyles.css" as="style" onload="this.rel='stylesheet'">
```

(Just to make it clear, I'm fine with changing it to `data-async`, as to me it has almost equal weight of pros and cons, just please consider the above. So we will not change our mind in future)

warpech replied:
When defining attributes for a custom element, of course it is fine to not have a `data-` prefix.

But if your JavaScript tampers with some random element it finds in DOM, it is good to leave a trace in code that answers the question "for whom does this attribute matter".

So maybe just change it to `starcounter-include-non-blocking`?

I like the `<link rel="preload" href="mystyles.css" as="style" onload="this.rel='stylesheet'">` solution, too.

tomalec replied:
I find `starcounter-include-non-blocking` verbose, what about `sc-non-blocking` which is shorter and does not tangle implementation details for app dev (he/she does not need to care that it is `..-include`s responsibility.

tomalec replied:
We agreed with @warpech on `<link rel="preload" href="mystyles.css" as="style" onload="this.rel='stylesheet'">` which is explicit, Starcounter-independent way to add non blocking `link` in any web page.  We will test only that we DO NOT wait for `link[rel='preload']`.

warpech commented on README.md:
```diff
@@ -143,6 +144,11 @@ Name                                    | Detail                 | Description
+## Render-blocking links
+
+To mimic native behavior of markup for Declarative Shadow DOM and to avoid <abbr title="Flash Of Unstyled Content">FOUC</abbr> the element will "block rendering" until all `<link rel="stylesheet">`s are loaded or throw an error.
+You can opt-out by using the `async` boolean attribute on your `<link>`, like `<link rel="stylesheet" href="path/to/style.css" async>`. The blocked rendering is achieved by setting `visibility: hidden` on shadow host - `<strcounter-include>` element.
```
These two sentences have little in common. I propose to move the sentence about the `async` attribute to a separate, last paragraph.

warpech commented on starcounter-include.html:
```diff
@@ -476,13 +519,25 @@
+            this.blockingShadowLinksLoaded && this.blockingShadowLinksLoaded.reject();
 
             if (givenComposition !== undefined) {
+                this.setAttribute('loading-presentation', '');
```
Can this be moved after the line 534? Then the else in that `if` won't be needed.

tomalec replied:
:+1: Great idea.

tomalec replied:
err...  after second thought, I'd say we still need the else. If the old composition is still being loaded. Then we stamp new one, which does not have any blocking links, we would like to remove attribute synchronously, not to have FONC (Flash Of No Content) - attribute removed on (even instantly) rejected promise will be removed asynchronously.

tomalec replied:
but still we can move `setAttr..` to l.534,  as this is the only case and time we need it.

warpech replied:
I agree

warpech commented on starcounter-include.html:
```diff
@@ -476,13 +519,25 @@
+                if(blockingShadowLinksLoaded){
+                    blockingShadowLinksLoaded.then(()=>{
+                        this.removeAttribute('loading-presentation');
+                    });
```
Shouldn't you also add the function that removes the attribute in case of rejection?

```js
                    blockingShadowLinksLoaded.then(()=>{
                        this.removeAttribute('loading-presentation');
                    }, ()=>{
                        this.removeAttribute('loading-presentation');
                    });
```

tomalec replied:
Please note, that `error` of `<link>` also resolves the promise.

The only rejection happens when the new composition is stamped (l.524). (or may happen if someday we will cal reject elsewhere)
This means that links for the given presentation were not loaded yet, meaning the content is still unstyled. Then why should we reveal it? The reason to reveal it will happen only once we load new composition with styled content.

warpech replied:
I thought that the attribute `loading-presentation` is used to indicate that we are waiting for presentation links to be loaded, and that it is removed when we are no longer waiting.

Otherwise, perhaps it would be good to rename this attribute to `unloaded-presentation`?

warpech replied:
@tomalec explained to me that the only time when you will have the promise rejected is when a new view is being loaded in place of another view that was not loaded yet. So handling this case has no benefit, let's leave the `loading-presentation` attribute as is.

tomalec committed:
Fix typo in code comment,

https://github.com/Starcounter/starcounter-include/pull/111#discussion_r215269185

tomalec committed:
Improve path resolution in mockserver,

https://github.com/Starcounter/starcounter-include/pull/111#discussion_r215271675
https://github.com/Starcounter/starcounter-include/pull/111#discussion_r215273089

tomalec committed:
Divide "Render-blocking links" README section into paragraphs,

https://github.com/Starcounter/starcounter-include/pull/111#discussion_r215593392

tomalec committed:
Set the `loading-presentation` style only when it's needed

https://github.com/Starcounter/starcounter-include/pull/111#discussion_r215607176

tomalec commented:
Addressed most of the comments.
I have more doubts to some of them, especially:
- [x] https://github.com/Starcounter/starcounter-include/pull/111#discussion_r215300613
- [x] https://github.com/Starcounter/starcounter-include/pull/111#discussion_r215270576
- [x] https://github.com/Starcounter/starcounter-include/pull/111#discussion_r215593192
- [x] https://github.com/Starcounter/starcounter-include/pull/111#discussion_r215607176
- [x] https://github.com/Starcounter/starcounter-include/pull/111#discussion_r215725315
- [x] https://github.com/Starcounter/starcounter-include/pull/111#discussion_r215730655

tomalec committed:
Improve code comments around `loading-presentation` attribute

warpech commented on starcounter-include.html:
```diff
@@ -476,13 +520,28 @@
         StarcounterIncludePrototype.stampComposition = function(givenComposition){
-            // create shadowRoot if none
             const shadowRoot = this.shadowRoot;
+            this.blockingShadowLinksLoaded && this.blockingShadowLinksLoaded.reject();
```
If I comment out this line, all tests still pass. What is the purpose of it?

tomalec replied:
In case when you change the composition before previous one is loaded, you don't want the old promise to resolve and remove the `loading` attribute once the old finish loading and while new one is still being loaded.

tomalec replied:
I'm trying to write a test for it.

tomalec committed:
Add `presentation-loaded` event

https://github.com/Starcounter/starcounter-include/pull/111#discussion_r215730655

tomalec committed:
Remove `async` attribute support,

promote `link rel="preload"` solution
https://github.com/Starcounter/starcounter-include/pull/111#discussion_r215593192

tomalec committed:
Add tests for loading new composition slower than the old one's load event,

Make error caching cleaner,
https://github.com/Starcounter/starcounter-include/pull/111#discussion_r216246667

warpech commented on README.md:
```diff
@@ -60,8 +60,9 @@ To produce
- - Easy way to attach presentation expressed in declarative Shadow DOM
+  - Supports `<script>, <style>` tags per template instance,
+ - Easy way to attach presentation expressed in declarative Shadow DOM,
+ - Blocks rendering of Shadow DOM until `<link rel="stylesheet">`s are loaded (unless marked with `async` attribute).
```
Unless loaded asynchrously (see below)

warpech commented on starcounter-include.html:
```diff
@@ -476,13 +520,36 @@
+                        this.removeAttribute('loading-presentation');
+                        this.dispatchEvent(new CustomEvent('presentation-loaded'));
+                    }, (reason) => {
+                        // forward uncought rejection
```
uncaught

tomalec committed:
Update `async` note in README,

https://github.com/Starcounter/starcounter-include/pull/111#discussion_r216989742
fix typo in code comment,
https://github.com/Starcounter/starcounter-include/pull/111#discussion_r216990029

tomalec committed:
Remove "already loaded" - `link.sheet` check

as it is unreliable in Edge, and newly created composition should load styles asynchronously

warpech reviewed (approved):
Lgtm, let's go with it!

tomalec commented:
Tested with https://github.com/Juicy/imported-template/pull/59 on UniformDocs, and works fine in Chrome, FF, Edge. I'll wait for `imported-template` to be released, as without it, it results in `loading-presentation` not being removed => blank page.

tomalec commented:
Maybe we should hard-code our own timeout in case of something else goes wrong with `load` events?

warpech commented:
I would skip that for now, unless we have a reproducible case that we can research on.

tomalec committed:
Bump the Juicy/imported-template@3.4.0

to emphasize the need for https://github.com/Juicy/imported-template/pull/59

tomalec committed:
Wait for actual link.load event instead of timeout,

to fix tests in Edge
